File: behaviour/State.php
<?php

//状态模式  状态切换

interface State
{
    function handle();
}

class FreeState implements State
{
    public function handle(){
        echo "空闲中\n";
    }
}

class UseState implements State
{
    public function handle(){
        echo "使用中\n";
    }
}

class Context
{
    private $state;

    public function setState(State $s)
    {
        echo "已更改\n";
        $this->state = $s;
        $this->state->handle();
    }

}

class Client
{
    function do()
    {
        $context = new Context();
        $context->setState(new FreeState());
        $context->setState(new UseState());
    }
}

$c = new Client();
$c->do();

File: ruby/behaviour/State.rb
module State
    
    class UsingState
        def initialize
            
        end
        def handle
            puts 'using state'
        end
    end
    class FreeState
        def initialize
            
        end
        def handle
            puts 'free state'
        end
    end

    class Context
        @state
        def initialize
            
        end
        def setState(state)
            @state=state
            puts 'handling state'
            @state.handle
        end
    end

    module Test

        fs = FreeState.new
        us = UsingState.new

        context = Context.new
        context.setState fs
        context.setState us
        
    end

end